api/python/quilt3/packages.py

```diff
@@ -583,18 +590,30 @@ def _browse(cls, name, registry=None, top_hash=None):
             registry.resolve_top_hash(name, top_hash)
         )
         pkg_manifest = registry.manifest_pk(name, top_hash)
-        if pkg_manifest.is_local():
-            local_pkg_manifest = pkg_manifest.path
-        else:
-            local_pkg_manifest = CACHE_PATH / "manifest" / _filesystem_safe_encode(str(pkg_manifest))
-            if not local_pkg_manifest.exists():
-                # Copy to a temporary file first, to make sure we don't cache a truncated file
-                # if the download gets interrupted.
-                tmp_path = local_pkg_manifest.with_suffix('.tmp')
-                copy_file(pkg_manifest, PhysicalKey.from_path(tmp_path), message="Downloading manifest")
-                tmp_path.rename(local_pkg_manifest)
-
-        return cls._from_path(local_pkg_manifest)
+
+        def download_manifest(dst):
+            copy_file(pkg_manifest, PhysicalKey.from_path(dst), message="Downloading manifest")
+
+        with contextlib.ExitStack() as stack:
+            if pkg_manifest.is_local():
+                local_pkg_manifest = pkg_manifest.path
+            elif util.IS_CACHE_ENABLED:
+                local_pkg_manifest = CACHE_PATH / "manifest" / _filesystem_safe_encode(str(pkg_manifest))
+                if not local_pkg_manifest.exists():
+                    # Copy to a temporary file first, to make sure we don't cache a truncated file
+                    # if the download gets interrupted.
+                    tmp_path = local_pkg_manifest.with_suffix('.tmp')
+                    download_manifest(tmp_path)
+                    tmp_path.rename(local_pkg_manifest)
+            else:
+                # This tmp file has to closed before downloading, because on Windows it can't be
+                # opened for concurrent access.
+                with tempfile.NamedTemporaryFile(delete=False) as tmp_file:
+                    local_pkg_manifest = tmp_file.name
+                    stack.callback(os.unlink, local_pkg_manifest)
```

Comment:
This seems a little bit over-engineered... Ideally, we'd just download the manifest and immediately deserialize it, without even writing to a temporary file - or create a temporary file, but use its file descriptor instead of opening it again. But I guess that's not supported right now.

Fine for now, but would be nice to clean up some day.